### DISCUSSION TITLE
Why is the structure factor calculation so slow for lennard-jones liquids at low densities?

### DISCUSSION THREAD

**Discussion by CreditDefaultSwap** [glotzerlab/freud#1249] [General]
Hello. I have been using Freud to calculate the structure factor for different values of temperature and density for a pair potential fitted to the Lennard-Jones. Specifically, I am loading trajectories from an *.gsd file and then using 
freud.diffraction.StaticStructureFactorDirect.compute(frame,reset=False).
This generally works reasonably "fast" for high densities, \rho, and temperatures, T, of (\rho>0.1 and \T>0.8, in reduced units) and the simulation, from kmin=0.2 to kmax=20 and nbins=100 (50k frames) only takes a few hours in a 256 thread machine. However, for low densities of \rho=0.05, suddenly the computation will take around 28 hr for otherwise the same number of particles and machine!

I understand the box size increases at lower density (trajectories are simulated in an NVT ensemble), but I fail to understand why the computation seems to take exponentially longer at  these densities when it's otherwise the same number of particles and same number of wave vector bins.  

I rather use an integrated package like Freud and not have to compute this value myself, so I appreciate any clarification on how to overcome this density dependence, especially if it's just a set-up issue. Thanks a lot for any help or comments.

**Comment by tommy-waltmann** [comment]
Hi @CreditDefaultSwap ,

The reason the `StaticStructureFactorDirect` code slows down at larger box sizes is because is has to sample vectors in k-space in an isotropic distribution, and the size of the k-space bins is inversely proportional to the length of the box vectors. The best advice I can give without explicitly seeing the system and script demonstrating the problem is to lower the `k_max` for the lower density system.

There is some more discussion of memory and time complexity of the direct structure factor on #1242 , in which lowering the `k_max` seems to have fixed the memory issue another user was experiencing.

**Reply by CreditDefaultSwap** [reply]
Hi Tommy,

I see, of course, makes sense. I had misunderstood how the algorithm was being implemented. Just to be safe, however, based on the discussion on #1242, Freud picks a set of $\sim O(k_{max}^3*(L_x * L_y * L_z))$ points right, not $\sim O(k_{max}^3/(L_x * L_y * L_z))$ considering that the smallest $k$ value is of the order $1/L$?
I appreciate the clarification.

**Reply by tommy-waltmann** [reply]
Yes, you're right. That was a typo in my original message.

**Reply by CreditDefaultSwap** [reply]
I see thank you very much. Was really helpful discussion.
